File: 323. Number of Connected Components in an Undirected Graph/323.cpp
// disjoint set union
// E = number of edges, V = number of vertices
// time: O(V + E * alpha(n)) space: O(V)
class Solution {
public:
    int findRep(vector<int> representative, int node){
        // the representative's node's representative would always be itself
        if (representative[node] == node){
            return node;
        }

        return findRep(representative, representative[node]);
    }

    int combine(int node1, int node2, vector<int>& connected_size, vector<int>& representative){
        int rep1 = findRep(representative, node1);
        int rep2 = findRep(representative, node2);
        
        // nodes belong in same connected component
        if (rep1 == rep2){
            return 0;
        } else if (connected_size[rep1] > connected_size[rep2]){
            connected_size[rep1] += connected_size[rep2];
            representative[rep2] = rep1;
            return 1;
        } else{
           connected_size[rep2] += connected_size[rep1];
            representative[rep1] = rep2;
            return 1;
        }
    }

    int countComponents(int n, vector<vector<int>>& edges) {
        vector<int> representative(n);
        vector<int> connected_size(n);
        
        // initialization
        for (int i = 0; i < n; i++){
            representative[i] = i;
            connected_size[i] = 1;
        }

        int numconnected = n;
        for(vector<int> edge: edges){
            numconnected -= combine(edge[0], edge[1], connected_size, representative);
        }

        return numconnected;
    }

};

File: 323. Number of Connected Components in an Undirected Graph/323.py
"""
E = Number of edges, V = Number of vertices
time: O(V+E⋅α(n))
space: O(V)
"""
class Solution:
    def getRepresentative(self, node, representative):
        if representative[node] == node:
            return node
        
        return self.getRepresentative(representative[node], representative)

    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        numConnected = n
        representative = {}

        for i in range(n):
            representative[i] = i

        for edge in edges:
            a = edge[0]
            b = edge[1] 

            a_rep = self.getRepresentative(a, representative)
            b_rep = self.getRepresentative(b, representative)

            if a_rep < b_rep:
                representative[b_rep] = a_rep
                numConnected -= 1
            elif b_rep < a_rep:
                representative[a_rep] = b_rep
                numConnected -= 1

        return numConnected


"""
representative_map = {node: representative}
numConnected
"""
